plugins/optimization/src/main/java/org/qbicc/plugin/opt/ea/EscapeAnalysisIntraMethodBuilder.java

```diff
@@ -0,0 +1,174 @@
+            isSubtypeOfClass("java/lang/ThreadGroup", type)) {
+            return EscapeValue.GLOBAL_ESCAPE;
+        }
+
+        return EscapeValue.NO_ESCAPE;
+    }
+
+    private boolean isSubtypeOfClass(String name, ClassObjectType type) {
+        return type.isSubtypeOf(bootstrapClassContext.findDefinedType(name).load().getType());
+    }
+
+    @Override
+    public ValueHandle instanceFieldOf(ValueHandle handle, FieldElement field) {
+        final InstanceFieldOf result = (InstanceFieldOf) super.instanceFieldOf(handle, field);
+
+        // T a = new T(...);
+        // To get represent the GC from 'a' to 'new T(...)',
+        // we hijack future 'a' references to fix the pointer.
+        // When 'a.x' is accessed, we fix the pointer from 'a' to 'new T(...)'.
+        handleInstanceFieldOf(result, handle, handle);
+
+        return result;
+    }
+
+    @Override
+    public Node store(ValueHandle handle, Value value, WriteAccessMode mode) {
+        final Node result = super.store(handle, value, mode);
+
+        if (handle instanceof StaticField) {
+            // static T a = ...
+            if (value instanceof NotNull nn) {
+                connectionGraph.trackStoreStaticField(handle, nn.getInput());
+            } else {
+                connectionGraph.trackStoreStaticField(handle, value);
+            }
+        } else if (handle instanceof InstanceFieldOf && value instanceof New) {
+            // p.f = new T(); // where p is a parameter
+            connectionGraph.fixEdgesNew(handle, (New) value);
+        } else if (handle instanceof LocalVariable && value instanceof New) {
+            connectionGraph.trackLocalNew((LocalVariable) handle, (New) value);
+        }
+
+        return result;
+    }
+
+    @Override
+    public Value call(ValueHandle target, List<Value> arguments) {
+        final Value result = super.call(target, arguments);
+
+        if (target instanceof Executable) {
+            escapeAnalysisState.trackCall(getCurrentElement(), (Call) result);
+        }
+
+        return result;
+    }
+
+    @Override
+    public void startMethod(List<ParameterValue> arguments) {
+        super.startMethod(arguments);
+        connectionGraph.trackParameters(arguments);
+    }
+
+    @Override
+    public BasicBlock return_(Value value) {
+        final BasicBlock result = super.return_(value);
+
+        // Skip primitive values truncated, they are not objects
+        if (!(value instanceof Truncate)) {
+            connectionGraph.trackReturn(value);
+        }
+
+        return result;
+    }
+
+    @Override
+    public BasicBlock throw_(Value value) {
+        final BasicBlock result = super.throw_(value);
+
+        if (value instanceof New) {
+            connectionGraph.trackThrowNew((New) value);
+        }
+
+        return result;
+    }
+
+    @Override
+    public Value checkcast(Value value, Value toType, Value toDimensions, CheckCast.CastType kind, ObjectType expectedType) {
```

Comment:
I think there are likely many kinds of casts you'll need to be able to see through.  `CheckCast` in particular can be optimized away or optimized down to `BitCast` as early as `ADD`.

Reply:
Ok. Do you have java code examples that show this situations in action?

Reply:
Not on hand but this situation should be easy enough to set up if you cast a reference to a more general type that it definitely already implements, like `Object`, I think.  Or maybe cast to `Object` and then back again.